Add xdotool & libxdo keyboard implementations

Danesprite opened the pull request:
Xdotool should be replacing the current pynput implementation used on Linux. A libxdo implementation has also been added, which is faster than xdotool, but is disabled because it doesn't work properly with Python 3.x. Both of these have been adapted from Aenea's implementations.

This should fix #79 from my testing. Keys are not marked as synthetic any more and keyboard shortcuts seem to be working correctly.

I would like to allow specifying other key symbols and characters through `Key` and `Text`. E.g. `Key("XF86Mail")` and `Text("€", is_text=True)`. Example 2 is already possible on Windows.

I am hoping to use more of Aenea's xdotool implementation for X11 `Mouse` and `Window` classes.

@shervinemami @dwks @GoNZooo @calmofthestorm You all may be interested in this :-)

shervinemami commented:
Nice! I went through a lot of pain when trying to come up with single-word fixed-font Unicode symbols to use for BeamGrid, eventually I found enough symbols that were supported by the font and Aenea and xdotool, but it was quite a pain: https://github.com/shervinemami/BeamGrid/blob/master/MacroSystem/symbolmap.py

Danesprite commented:
Oh cool! I'll have to give that a shot sometime. I still just move the mouse manually, but IMO trackballs aren't anywhere near as bad as regular mice RSI-wise.

Xdotool's  `type` command seems to type these characters properly for me. Just need to get `Text` to use it somehow.

For other keys like `XF86Mail` (and there are quite a few!), it should be possible to adjust the `Key` class to pass through keys that aren't in the `typeables` dictionary instead of raising an error. Xdotool will then handle keys that are actually invalid. I'll make it so this only happens on X11. Windows doesn't allow lookup of virtual keys like that AFAIK.

shervinemami commented:
Great, cos yeah the list I made is of symbols that are supported by typeables. But hopefully with your change, more keys and symbols will be supported. I'm currently sending a lot of commands between Windows & Linux by using ugly hotkey combinations such as Meta+Shift+X, but I've already run out of hotkeys I'm willing to use, and it would be much nicer if we could use the more than 100 other unsupported keys like mentioned in the "Keyboard" or other sections of "http://www.freebsddiary.org/APC/usb_hid_usages.php".

Danesprite commented:
Ah, yeah that really doesn't sound great. Once I have implemented this it should be possible to type any key that xdotool supports without using hotkey combinations like that. So a lot of those keys will be supported. I think Aenea would need to be changed to allow it though.

I figured out that you can specify the Unicode code point to get `xdotool key` to type symbols like this:

```Python
>>> u"€".encode("unicode_escape")[1:].upper()
'U20AC'

```

Then you can do `xdotool key U20AC` to type "€".

shervinemami commented:
Nice!

dwks commented:
Haven't tested it, but it looks good. Thanks for getting around to this!

Danesprite commented:
@dwks No worries. Sorry it took me so long.

I've implemented the changes mentioned above:

-  `Key` and `Text` actions should be able to type any Unicode characters that xdotool can (e.g. using `Key("€,€")` or `Text("€€")`). This also works perfectly with the libxdo implementation in Python 2.7. The `is_text` parameter is not required for either action. You can even type emoji with it!

- The `Key` action can also now type any valid xdotool key (e.g. `XF86Refresh`, `XF86Mail`, `U3041`, etc). If the key is invalid, xdotool will print `(symbol) No such key name` warnings.

I've updated the `dragonfly/actions/_test_x11_text_key.py` test script with new characters. Let me know what you all think! :-)

I need to mention somewhere in the documentation that xdotool must be installed for this to work. Maybe the libxdo implementation could be enabled via a setting under `~/.dragonfly2-speech/`.

calmofthestorm commented:
Neat! Never thought we'd still be shelling out to xdotool after all these years, but I'm glad it gets the job done!

Danesprite commented:
@calmofthestorm Yep! Xdotool really seems to still be the best & easiest way of doing this.

I'll be adding an X11 version of dragonfly's `Window` class soon. It's based on your work in Aenea using xdotool and xprop, so thank you for that! :-)
